File: plugins/kinetica/skills/kinetica-execute/scripts/modules/viz.js
'use strict';

/**
 * Visualization module for the Kinetica CLI (Node.js).
 *
 * Provides 5 commands: chart, heatmap, isochrone, classbreak, wms.
 *
 * Each export follows the category module contract:
 *   { fn: async (db, args) => void, desc: string }
 */

const fs = require('fs');
const { die, out, parseCsvArg } = require('./helpers');

// ---------------------------------------------------------------------------
// Image output helpers
// ---------------------------------------------------------------------------

/**
 * Write image data to file or report its length.
 * Auto-detects whether image_data is raw binary (PNG header) or base64-encoded.
 * @param {object} resp - Kinetica API response containing image_data
 * @param {string|undefined} outputPath - File path to write the image to
 */
function handleImageOutput(resp, outputPath, previewWidth) {
  const imageData = resp.image_data || '';

  // Decode image data to a buffer for both preview and file output
  const isRawBinary = imageData.length >= 4 &&
    imageData.charCodeAt(0) === 0x89 &&
    imageData.slice(1, 4) === 'PNG';
  const decoded = isRawBinary
    ? Buffer.from(imageData, 'binary')
    : Buffer.from(imageData, 'base64');

  if (previewWidth) {
    const { renderPreview } = require('./image-preview');
    renderPreview(decoded, { maxWidth: previewWidth > 1 ? previewWidth : 0 });
  }

  if (outputPath) {
    fs.writeFileSync(outputPath, decoded);
    out({
      status: 'ok',
      output: outputPath,
      size_bytes: decoded.length,
    });
  } else {
    out({
      status: 'ok',
      image_data_length: imageData.length,
    });
  }
}

/**
 * Write raw binary image data to file or report its size.
 * Used by WMS-based commands (heatmap, classbreak, wms).
 * @param {Buffer} buffer - Raw PNG bytes from WMS
 * @param {string|undefined} outputPath - File path to write the image to
 */
function handleBinaryImageOutput(buffer, outputPath, previewWidth) {
  if (previewWidth) {
    const { renderPreview } = require('./image-preview');
    renderPreview(buffer, { maxWidth: previewWidth > 1 ? previewWidth : 0 });
  }

  if (outputPath) {
    fs.writeFileSync(outputPath, buffer);
    out({
      status: 'ok',
      output: outputPath,
      size_bytes: buffer.length,
    });
  } else {
    out({
      status: 'ok',
      size_bytes: buffer.length,
    });
  }
}

// ---------------------------------------------------------------------------
// WMS parameter builder
// ---------------------------------------------------------------------------

/**
 * Build a flat WMS parameter object with sensible defaults.
 * @param {object} opts - Structured options
 * @param {string} opts.table - LAYERS value (table name)
 * @param {string} [opts.styles] - WMS STYLES value (e.g. 'heatmap', 'cb_raster')
 * @param {string} [opts.srs='EPSG:4326'] - Spatial reference system
 * @param {number} [opts.minX=-180] - Bounding box min X
 * @param {number} [opts.minY=-90] - Bounding box min Y
 * @param {number} [opts.maxX=180] - Bounding box max X
 * @param {number} [opts.maxY=90] - Bounding box max Y
 * @param {number} [opts.width=800] - Image width
 * @param {number} [opts.height=600] - Image height
 * @param {string} [opts.xAttr] - X attribute column (mutually exclusive with geoAttr)
 * @param {string} [opts.yAttr] - Y attribute column (mutually exclusive with geoAttr)
 * @param {string} [opts.geoAttr] - WKT/geometry column (mutually exclusive with xAttr/yAttr)
 * @param {object} [opts.extra={}] - Additional WMS params merged last
 * @returns {object} Flat WMS parameter dict
 */
function buildWmsParams(opts) {
  if (opts.geoAttr && (opts.xAttr || opts.yAttr)) {
    die('GEO_ATTR and X_ATTR/Y_ATTR are mutually exclusive');
  }

  const WORLD = { minX: -180, minY: -90, maxX: 180, maxY: 90 };
  const PAD = 0.3;

  let minX = opts.minX != null ? opts.minX : WORLD.minX;
  let minY = opts.minY != null ? opts.minY : WORLD.minY;
  let maxX = opts.maxX != null ? opts.maxX : WORLD.maxX;
  let maxY = opts.maxY != null ? opts.maxY : WORLD.maxY;

  const isCustom = minX !== WORLD.minX || minY !== WORLD.minY
                || maxX !== WORLD.maxX || maxY !== WORLD.maxY;
  if (isCustom) {
    minX = Math.max(WORLD.minX, minX - PAD);
    minY = Math.max(WORLD.minY, minY - PAD);
    maxX = Math.min(WORLD.maxX, maxX + PAD);
    maxY = Math.min(WORLD.maxY, maxY + PAD);
  }

  const params = {
    REQUEST: 'GetMap',
    FORMAT: 'image/png',
    SRS: opts.srs || 'EPSG:4326',
    LAYERS: opts.table,
    BBOX: [minX, minY, maxX, maxY].join(','),
    WIDTH: opts.width || 800,
    HEIGHT: opts.height || 600,
  };

  if (opts.styles) params.STYLES = opts.styles;
  if (opts.geoAttr) {
    params.GEO_ATTR = opts.geoAttr;
  } else {
    if (opts.xAttr) params.X_ATTR = opts.xAttr;
    if (opts.yAttr) params.Y_ATTR = opts.yAttr;
  }

  return { ...params, ...(opts.extra || {}) };
}

// ---------------------------------------------------------------------------
// Config file loader
// ---------------------------------------------------------------------------

/**
 * Load a JSON config from a @file path or inline string.
 * @param {string} configArg - Either "@path/to/file.json" or inline JSON
 * @returns {object} Parsed config object
 */
function loadConfig(configArg) {
  if (!configArg) {
    die('--config is required (e.g. --config @file.json or inline JSON)');
  }

  if (configArg.startsWith('@')) {
    const filePath = configArg.slice(1);
    if (!fs.existsSync(filePath)) {
      die(`Config file not found: ${filePath}`);
    }
    try {
      return JSON.parse(fs.readFileSync(filePath, 'utf8'));
    } catch (err) {
      die(`Invalid JSON in config file ${filePath}: ${err.message}`);
    }
  }

  try {
    return JSON.parse(configArg);
  } catch (err) {
    die(`Invalid inline JSON config: ${err.message}`);
  }

  // Unreachable, but keeps linters happy
  return {};
}

// ---------------------------------------------------------------------------
// chart
// ---------------------------------------------------------------------------

async function cmdChart(db, args) {
  const tableName = args.positional[0];
  if (!tableName) {
    die(
      'Usage: viz chart <table> --x-column COL --y-column COL ' +
        '[--point-color RRGGBB] [--point-size N] [--point-shape circle|square|diamond] ' +
        '[--min-x N --max-x N --min-y N --max-y N] ' +
        '[--width 800] [--height 600] [--bg-color FFFFFF] [--output file.png]'
    );
  }

  const xColumn = args.flags['x-column'];
  const yColumn = args.flags['y-column'];
  if (!xColumn || !yColumn) {
    die('--x-column and --y-column are required');
  }

  const minX = parseFloat(args.flags['min-x'] || '0');
  const maxX = parseFloat(args.flags['max-x'] || '0');
  const minY = parseFloat(args.flags['min-y'] || '0');
  const maxY = parseFloat(args.flags['max-y'] || '0');
  const width = parseInt(args.flags.width || '800', 10);
  const height = parseInt(args.flags.height || '600', 10);
  const bgColor = args.flags['bg-color'] || 'FFFFFF';

  const styleOptions = {};
  if (args.flags['point-color']) styleOptions.pointcolor = [args.flags['point-color']];
  if (args.flags['point-size']) styleOptions.pointsize = [args.flags['point-size']];
  if (args.flags['point-shape']) styleOptions.pointshape = [args.flags['point-shape']];

  const resp = await db.visualize_image_chart(
    tableName,
    [xColumn],
    [yColumn],
    minX,
    maxX,
    minY,
    maxY,
    width,
    height,
    bgColor,
    styleOptions,
    {}
  );

  const previewWidth = args.flags.preview ? parseInt(args.flags['preview-width'] || '0', 10) || true : 0;
  handleImageOutput(resp, args.flags.output, previewWidth);
}

// ---------------------------------------------------------------------------
// heatmap (WMS-based)
// ---------------------------------------------------------------------------

async function cmdHeatmap(db, args) {
  const tableName = args.positional[0];
  if (!tableName) {
    die(
      'Usage: viz heatmap <table> (--x-col COL --y-col COL | --geo-col COL) ' +
        '[--value-col COL] [--srs EPSG:4326] [--blur-radius 5] [--colormap jet] ' +
        '[--min-x N --max-x N --min-y N --max-y N] ' +
        '[--width 800] [--height 600] [--output file.png]'
    );
  }

  const xCol = args.flags['x-col'];
  const yCol = args.flags['y-col'];
  const geoCol = args.flags['geo-col'];

  if (geoCol && (xCol || yCol)) {
    die('--geo-col and --x-col/--y-col are mutually exclusive');
  }
  if (!geoCol && (!xCol || !yCol)) {
    die('Either --geo-col or both --x-col and --y-col are required');
  }

  const extra = {};
  const valueCol = args.flags['value-col'];
  if (valueCol) extra.VALUE_ATTR = valueCol;

  const blurRadius = args.flags['blur-radius'];
  if (blurRadius) extra.BLUR_RADIUS = blurRadius;

  const colormap = args.flags.colormap;
  if (colormap) extra.COLORMAP = colormap;

  const params = buildWmsParams({
    table: tableName,
    styles: 'heatmap',
    srs: args.flags.srs || 'EPSG:4326',
    minX: parseFloat(args.flags['min-x'] || '-180'),
    minY: parseFloat(args.flags['min-y'] || '-90'),
    maxX: parseFloat(args.flags['max-x'] || '180'),
    maxY: parseFloat(args.flags['max-y'] || '90'),
    width: parseInt(args.flags.width || '800', 10),
    height: parseInt(args.flags.height || '600', 10),
    xAttr: xCol,
    yAttr: yCol,
    geoAttr: geoCol,
    extra,
  });

  const buffer = await db.wms_request(params);
  const previewWidth = args.flags.preview ? parseInt(args.flags['preview-width'] || '0', 10) || true : 0;
  handleBinaryImageOutput(buffer, args.flags.output, previewWidth);
}

// ---------------------------------------------------------------------------
// isochrone
// ---------------------------------------------------------------------------

async function cmdIsochrone(db, args) {
  const graphName = args.positional[0];
  if (!graphName) {
    die(
      'Usage: viz isochrone <graph> --source NODE ' +
        '[--max-radius N] [--num-levels N] [--levels-table NAME] [--output file.png]'
    );
  }

  const sourceNode = args.flags.source;
  if (!sourceNode) {
    die('--source is required');
  }

  const maxRadius = parseFloat(args.flags['max-radius'] || '100');
  const numLevels = parseInt(args.flags['num-levels'] || '4', 10);
  const levelsTable = args.flags['levels-table'] || '';
  const weightsOnEdges = parseCsvArg(args.flags['weights-on-edges']);
  const restrictions = parseCsvArg(args.flags.restrictions);
  const outputPath = args.flags.output;

  // Generate image only when --output is provided or no levels-table
  const generateImage = Boolean(outputPath) || !levelsTable;

  const resp = await db.visualize_isochrone(
    graphName,
    sourceNode,
    maxRadius,
    weightsOnEdges,
    restrictions,
    numLevels,
    generateImage,
    levelsTable,
    {},
    {},
    {},
    {}
  );

  if (generateImage) {
    const previewWidth = args.flags.preview ? parseInt(args.flags['preview-width'] || '0', 10) || true : 0;
    handleImageOutput(resp, outputPath, previewWidth);
  } else {
    out({
      status: 'ok',
      levels_table: levelsTable,
    });
  }
}

// ---------------------------------------------------------------------------
// classbreak (WMS-based)
// ---------------------------------------------------------------------------

function buildClassbreakParams(config) {
  const base = buildWmsParams({
    table: config.table || config.LAYERS,
    styles: 'cb_raster',
    srs: config.srs || config.SRS,
    minX: config.min_x,
    minY: config.min_y,
    maxX: config.max_x,
    maxY: config.max_y,
    width: config.width || config.WIDTH,
    height: config.height || config.HEIGHT,
    xAttr: config.x_attr,
    yAttr: config.y_attr,
    geoAttr: config.geo_attr,
    extra: config.BBOX ? { BBOX: config.BBOX } : {},
  });

  if (!base.LAYERS) die('Config must include "LAYERS" or "table"');

  // Pass through uppercase WMS keys not already set
  const passthrough = {};
  for (const [key, val] of Object.entries(config)) {
    if (key === key.toUpperCase() && !base[key]) passthrough[key] = val;
  }
  return { ...base, ...passthrough };
}

async function cmdClassbreak(db, args) {
  const configArg = args.flags.config;
  const config = loadConfig(configArg);
  const outputPath = args.flags.output;
  const previewWidth = args.flags.preview ? parseInt(args.flags['preview-width'] || '0', 10) || true : 0;

  const params = buildClassbreakParams(config);
  const buffer = await db.wms_request(params);
  handleBinaryImageOutput(buffer, outputPath, previewWidth);
}

// ---------------------------------------------------------------------------
// wms (general-purpose WMS command)
// ---------------------------------------------------------------------------

async function cmdWms(db, args) {
  const configArg = args.flags.config;
  const config = loadConfig(configArg);
  const outputPath = args.flags.output;
  const previewWidth = args.flags.preview ? parseInt(args.flags['preview-width'] || '0', 10) || true : 0;

  // Apply defaults, then merge user config
  const params = {
    REQUEST: 'GetMap',
    FORMAT: 'image/png',
    SRS: 'EPSG:4326',
    WIDTH: 800,
    HEIGHT: 600,
    ...config,
  };

  if (!params.LAYERS) {
    die('Config must include "LAYERS" (table name)');
  }
  if (!params.BBOX) {
    die('Config must include "BBOX" (e.g. "-180,-90,180,90")');
  }

  const buffer = await db.wms_request(params);
  handleBinaryImageOutput(buffer, outputPath, previewWidth);
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

module.exports = {
  chart: {
    fn: cmdChart,
    desc: 'Generate a chart image (line, bar, scatter, etc.)',
  },
  heatmap: {
    fn: cmdHeatmap,
    desc: 'Generate a heatmap image via WMS',
  },
  isochrone: {
    fn: cmdIsochrone,
    desc: 'Generate isochrone contours from a graph',
  },
  classbreak: {
    fn: cmdClassbreak,
    desc: 'Generate a class-break visualization via WMS',
  },
  wms: {
    fn: cmdWms,
    desc: 'Send a custom WMS request and save the image',
  },
};

File: plugins/kinetica/skills/kinetica-execute/scripts/modules/viz.py
"""Visualization module for the Kinetica CLI (Python).

Provides 5 commands: chart, heatmap, isochrone, classbreak, wms.

Each command follows the category module contract:
    COMMANDS = { "name": {"fn": callable, "desc": str, "build_args": callable} }
"""

import base64
import json
import urllib.error
import urllib.parse
import urllib.request

from modules.helpers import build_auth_headers, check_status, die, env, out


CHART_TYPES = [
    "line",
    "bar",
    "scatter",
    "area",
    "stacked_bar",
    "stacked_area",
]


# ---------------------------------------------------------------------------
# Image output helpers
# ---------------------------------------------------------------------------

def _handle_image_output(resp, output_path):
    """Write image data to file or report its length.

    Auto-detects whether image_data is raw binary (PNG header) or base64-encoded.
    """
    image_data = resp.get("image_data", "")
    if output_path:
        # Detect raw binary PNG (starts with \x89PNG) vs base64
        if isinstance(image_data, bytes):
            decoded = image_data
        elif len(image_data) >= 4 and image_data[:4] == '\x89PNG':
            decoded = image_data.encode("latin-1")
        else:
            decoded = base64.b64decode(image_data)
        with open(output_path, "wb") as f:
            f.write(decoded)
        out({
            "status": "ok",
            "output": output_path,
            "size_bytes": len(decoded),
        })
    else:
        out({
            "status": "ok",
            "image_data_length": len(image_data),
        })


def _handle_binary_image_output(data, output_path):
    """Write raw binary image data to file or report its size.

    Used by WMS-based commands (heatmap, classbreak, wms).
    """
    if output_path:
        with open(output_path, "wb") as f:
            f.write(data)
        out({
            "status": "ok",
            "output": output_path,
            "size_bytes": len(data),
        })
    else:
        out({
            "status": "ok",
            "size_bytes": len(data),
        })


# ---------------------------------------------------------------------------
# WMS parameter builder
# ---------------------------------------------------------------------------

def _build_wms_params(opts):
    """Build a flat WMS parameter dict with sensible defaults.

    Args:
        opts: dict with keys table, styles, srs, min_x, min_y, max_x, max_y,
              width, height, x_attr, y_attr, extra.
    Returns:
        Flat dict of WMS query parameters.
    """
    min_x = opts.get("min_x", -180)
    min_y = opts.get("min_y", -90)
    max_x = opts.get("max_x", 180)
    max_y = opts.get("max_y", 90)

    params = {
        "REQUEST": "GetMap",
        "FORMAT": "image/png",
        "SRS": opts.get("srs", "EPSG:4326"),
        "LAYERS": opts["table"],
        "BBOX": f"{min_x},{min_y},{max_x},{max_y}",
        "WIDTH": str(opts.get("width", 800)),
        "HEIGHT": str(opts.get("height", 600)),
    }

    if opts.get("styles"):
        params["STYLES"] = opts["styles"]
    if opts.get("x_attr"):
        params["X_ATTR"] = opts["x_attr"]
    if opts.get("y_attr"):
        params["Y_ATTR"] = opts["y_attr"]

    extra = opts.get("extra", {})
    return {**params, **extra}


# ---------------------------------------------------------------------------
# WMS HTTP request (Python has no SDK method for this)
# ---------------------------------------------------------------------------

def _wms_request(params):
    """Send a WMS request to Kinetica and return raw PNG bytes.

    Reads connection info from environment variables (same as helpers.connect).
    """
    base_url = env("KINETICA_DB_SKILL_URL")
    if not base_url:
        die("KINETICA_DB_SKILL_URL is not set")

    query_string = urllib.parse.urlencode(params)
    url = f"{base_url.rstrip('/')}/wms?{query_string}"

    req = urllib.request.Request(url)

    for key, val in build_auth_headers().items():
        req.add_header(key, val)

    try:
        with urllib.request.urlopen(req) as resp:
            return resp.read()
    except urllib.error.HTTPError as exc:
        try:
            body = exc.read().decode("utf-8", errors="replace")
        except Exception:
            body = "(unreadable)"
        die(f"WMS request failed (HTTP {exc.code}): {body}")
    except urllib.error.URLError as exc:
        die(f"WMS request failed: {exc.reason}")


# ---------------------------------------------------------------------------
# Config file loader
# ---------------------------------------------------------------------------

def _load_config(config_arg):
    """Load a JSON config from a @file path or inline string.

    Returns the parsed dict.  Dies on parse/read errors.
    """
    if not config_arg:
        die("--config is required (e.g. --config @file.json or inline JSON)")

    if config_arg.startswith("@"):
        file_path = config_arg[1:]
        try:
            with open(file_path) as f:
                return json.load(f)
        except FileNotFoundError:
            die(f"Config file not found: {file_path}")
        except json.JSONDecodeError as exc:
            die(f"Invalid JSON in config file {file_path}: {exc}")
    else:
        try:
            return json.loads(config_arg)
        except json.JSONDecodeError as exc:
            die(f"Invalid inline JSON config: {exc}")

    # Unreachable, but keeps linters happy
    return {}


# ---------------------------------------------------------------------------
# chart
# ---------------------------------------------------------------------------

def _build_chart_args(parser):
    parser.add_argument("table_name", help="Source table name")
    parser.add_argument(
        "--x-column", dest="x_column", required=True,
        help="Column name for the X axis",
    )
    parser.add_argument(
        "--y-column", dest="y_column", required=True,
        help="Column name for the Y axis",
    )
    parser.add_argument(
        "--type", dest="chart_type", default="line", choices=CHART_TYPES,
        help="Chart type (default: line)",
    )
    parser.add_argument("--min-x", dest="min_x", type=float, default=0, help="Minimum X value")
    parser.add_argument("--max-x", dest="max_x", type=float, default=0, help="Maximum X value")
    parser.add_argument("--min-y", dest="min_y", type=float, default=0, help="Minimum Y value")
    parser.add_argument("--max-y", dest="max_y", type=float, default=0, help="Maximum Y value")
    parser.add_argument("--width", type=int, default=800, help="Image width in pixels (default: 800)")
    parser.add_argument("--height", type=int, default=600, help="Image height in pixels (default: 600)")
    parser.add_argument("--bg-color", dest="bg_color", default="FFFFFF", help="Background color hex (default: FFFFFF)")
    parser.add_argument("--output", default=None, help="Output file path for the image")


def cmd_chart(db, args):
    """Generate a chart image from table data."""
    table_name = args.table_name
    if not table_name:
        die("Usage: viz chart <table> --x-column COL --y-column COL [options]")

    style_options = {
        "chart_type": args.chart_type,
    }

    resp = db.visualize_image_chart(
        table_name=table_name,
        x_column_names=[args.x_column],
        y_column_names=[args.y_column],
        min_x=args.min_x,
        max_x=args.max_x,
        min_y=args.min_y,
        max_y=args.max_y,
        width=args.width,
        height=args.height,
        bg_color=args.bg_color,
        style_options=style_options,
        options={},
    )
    check_status(resp, "visualize_image_chart")
    _handle_image_output(resp, args.output)


# ---------------------------------------------------------------------------
# heatmap
# ---------------------------------------------------------------------------

def _build_heatmap_args(parser):
    parser.add_argument("table_name", help="Source table name")
    parser.add_argument(
        "--x-col", dest="x_col", required=True,
        help="Column name for X coordinate",
    )
    parser.add_argument(
        "--y-col", dest="y_col", required=True,
        help="Column name for Y coordinate",
    )
    parser.add_argument("--value-col", dest="value_col", default="", help="Column name for the heatmap value")
    parser.add_argument("--srs", default="EPSG:4326", help="Spatial reference system (default: EPSG:4326)")
    parser.add_argument("--blur-radius", dest="blur_radius", type=int, default=None, help="Blur radius (default: server default)")
    parser.add_argument("--colormap", default=None, help="Colormap name (e.g. jet, hot, viridis)")
    parser.add_argument("--min-x", dest="min_x", type=float, default=-180.0, help="Minimum X (default: -180)")
    parser.add_argument("--max-x", dest="max_x", type=float, default=180.0, help="Maximum X (default: 180)")
    parser.add_argument("--min-y", dest="min_y", type=float, default=-90.0, help="Minimum Y (default: -90)")
    parser.add_argument("--max-y", dest="max_y", type=float, default=90.0, help="Maximum Y (default: 90)")
    parser.add_argument("--width", type=int, default=800, help="Image width in pixels (default: 800)")
    parser.add_argument("--height", type=int, default=600, help="Image height in pixels (default: 600)")
    parser.add_argument("--output", default=None, help="Output file path for the image")


def cmd_heatmap(db, args):
    """Generate a heatmap image via WMS."""
    table_name = args.table_name
    if not table_name:
        die("Usage: viz heatmap <table> --x-col COL --y-col COL [options]")

    extra = {}
    if args.value_col:
        extra["VALUE_ATTR"] = args.value_col
    if args.blur_radius is not None:
        extra["BLUR_RADIUS"] = str(args.blur_radius)
    if args.colormap:
        extra["COLORMAP"] = args.colormap

    params = _build_wms_params({
        "table": table_name,
        "styles": "heatmap",
        "srs": args.srs,
        "min_x": args.min_x,
        "min_y": args.min_y,
        "max_x": args.max_x,
        "max_y": args.max_y,
        "width": args.width,
        "height": args.height,
        "x_attr": args.x_col,
        "y_attr": args.y_col,
        "extra": extra,
    })

    data = _wms_request(params)
    _handle_binary_image_output(data, args.output)


# ---------------------------------------------------------------------------
# isochrone
# ---------------------------------------------------------------------------

def _build_isochrone_args(parser):
    parser.add_argument("graph_name", help="Name of the graph")
    parser.add_argument(
        "--source", dest="source_node", required=True,
        help="Source node identifier",
    )
    parser.add_argument(
        "--max-radius", dest="max_radius", type=float, default=100.0,
        help="Maximum solution radius (default: 100)",
    )
    parser.add_argument("--weights-on-edges", dest="weights_on_edges", default="", help="Comma-separated weight identifiers")
    parser.add_argument("--restrictions", default="", help="Comma-separated restriction identifiers")
    parser.add_argument(
        "--num-levels", dest="num_levels", type=int, default=4,
        help="Number of contour levels (default: 4)",
    )
    parser.add_argument("--levels-table", dest="levels_table", default="", help="Output table name for contour levels")
    parser.add_argument("--output", default=None, help="Output file path for the image")


def cmd_isochrone(db, args):
    """Generate an isochrone contour from a graph."""
    graph_name = args.graph_name
    if not graph_name:
        die("Usage: viz isochrone <graph> --source NODE [options]")

    # Parse CSV lists
    weights = [w.strip() for w in args.weights_on_edges.split(",") if w.strip()] if args.weights_on_edges else []
    restrictions = [r.strip() for r in args.restrictions.split(",") if r.strip()] if args.restrictions else []

    # Generate image only when --output is provided or no levels-table
    generate_image = bool(args.output) or not args.levels_table

    resp = db.visualize_isochrone(
        graph_name=graph_name,
        source_node=args.source_node,
        max_solution_radius=args.max_radius,
        weights_on_edges=weights,
        restrictions=restrictions,
        num_levels=args.num_levels,
        generate_image=generate_image,
        levels_table=args.levels_table,
        style_options={},
        solve_options={},
        contour_options={},
        options={},
    )
    check_status(resp, "visualize_isochrone")

    if generate_image:
        _handle_image_output(resp, args.output)
    else:
        out({
            "status": "ok",
            "levels_table": args.levels_table,
        })


# ---------------------------------------------------------------------------
# classbreak
# ---------------------------------------------------------------------------

def _build_classbreak_args(parser):
    parser.add_argument(
        "--config", dest="config", required=True,
        help="JSON config: @file.json or inline JSON string",
    )
    parser.add_argument("--output", default=None, help="Output file path for the image")


def _build_classbreak_params(config):
    """Map a classbreak JSON config to flat WMS params.

    Delegates to _build_wms_params for core WMS parameters, then adds
    classbreak-specific overrides and uppercase key passthrough.
    """
    extra = {}
    bbox_val = config.get("BBOX")
    if bbox_val is not None:
        if isinstance(bbox_val, (list, tuple)):
            bbox_val = ",".join(str(v) for v in bbox_val)
        extra["BBOX"] = str(bbox_val)

    base = _build_wms_params({
        "table": config.get("table") or config.get("LAYERS", ""),
        "styles": "cb_raster",
        "srs": config.get("srs") or config.get("SRS") or "EPSG:4326",
        "min_x": config.get("min_x", -180),
        "min_y": config.get("min_y", -90),
        "max_x": config.get("max_x", 180),
        "max_y": config.get("max_y", 90),
        "width": config.get("width") or config.get("WIDTH") or 800,
        "height": config.get("height") or config.get("HEIGHT") or 600,
        "x_attr": config.get("x_attr"),
        "y_attr": config.get("y_attr"),
        "extra": extra,
    })

    if not base["LAYERS"]:
        die('Config must include "LAYERS" or "table"')

    # Pass through uppercase WMS keys not already set
    passthrough = {}
    for key, val in config.items():
        if key == key.upper() and key not in base:
            passthrough[key] = str(val) if not isinstance(val, str) else val
    return {**base, **passthrough}


def cmd_classbreak(db, args):
    """Generate a class-break visualization via WMS."""
    config = _load_config(args.config)
    output_path = args.output

    params = _build_classbreak_params(config)
    data = _wms_request(params)
    _handle_binary_image_output(data, output_path)


# ---------------------------------------------------------------------------
# wms (general-purpose WMS command)
# ---------------------------------------------------------------------------

def _build_wms_args(parser):
    parser.add_argument(
        "--config", dest="config", required=True,
        help="JSON config: @file.json or inline JSON string with WMS params",
    )
    parser.add_argument("--output", default=None, help="Output file path for the image")


def cmd_wms(db, args):
    """Send a custom WMS request and save the image."""
    config = _load_config(args.config)
    output_path = args.output

    # Apply defaults, then merge user config
    params = {
        "REQUEST": "GetMap",
        "FORMAT": "image/png",
        "SRS": "EPSG:4326",
        "WIDTH": "800",
        "HEIGHT": "600",
    }
    for key, val in config.items():
        params[key] = str(val) if not isinstance(val, str) else val

    if not params.get("LAYERS"):
        die('Config must include "LAYERS" (table name)')
    if not params.get("BBOX"):
        die('Config must include "BBOX" (e.g. "-180,-90,180,90")')

    data = _wms_request(params)
    _handle_binary_image_output(data, output_path)


# ---------------------------------------------------------------------------
# COMMANDS dict
# ---------------------------------------------------------------------------

COMMANDS = {
    "chart": {
        "fn": cmd_chart,
        "desc": "Generate a chart image (line, bar, scatter, etc.)",
        "build_args": _build_chart_args,
    },
    "heatmap": {
        "fn": cmd_heatmap,
        "desc": "Generate a heatmap image via WMS",
        "build_args": _build_heatmap_args,
    },
    "isochrone": {
        "fn": cmd_isochrone,
        "desc": "Generate isochrone contours from a graph",
        "build_args": _build_isochrone_args,
    },
    "classbreak": {
        "fn": cmd_classbreak,
        "desc": "Generate a class-break visualization via WMS",
        "build_args": _build_classbreak_args,
    },
    "wms": {
        "fn": cmd_wms,
        "desc": "Send a custom WMS request and save the image",
        "build_args": _build_wms_args,
    },
}
